File: FIRM-Core/FIRM_dsl/dynamic_evolution.py
"""dynamic_evolution.py

Dynamic phase evolution framework for Soul Garbage Collection.

Implements the theoretical equation:
dΦ_i/dt = -α_i ∇_Φ D_i + β_i Transmute(D_i) + γ_i Grace(Φ_i)

This bridges the gap between static analysis and dynamic evolution in the 𝒮-GC system.
"""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Dict, List, Tuple, Optional, Any
from enum import Enum
import math
import time

from .core import ObjectG, NodeLabel, normalize_phase_qpi
from .resonance import OmegaSignature, compute_resonance_alignment, derive_omega_signature
from .coherence import compute_coherence
from .grace_field import GraceFieldParams, potential_V, dV_du, FieldRegime


class EvolutionState(Enum):
    """States of dynamic evolution process."""
    STABLE = "stable"
    EVOLVING = "evolving"
    CONVERGING = "converging"
    DIVERGING = "diverging"
    OSCILLATING = "oscillating"


@dataclass
class EvolutionMetrics:
    """Metrics for tracking evolution progress."""
    coherence_history: List[float] = field(default_factory=list)
    resonance_history: List[float] = field(default_factory=list)
    phase_history: List[Dict[str, float]] = field(default_factory=list)
    evolution_rate: float = 0.0
    convergence_rate: float = 0.0
    oscillation_amplitude: float = 0.0
    state: EvolutionState = EvolutionState.STABLE


@dataclass
class DynamicPhaseEvolution:
    """Implements dynamic phase evolution for 𝒮-GC.

    Theoretical foundation:
    dΦ_i/dt = -α_i ∇_Φ D_i + β_i Transmute(D_i) + γ_i Grace(Φ_i)

    Where:
    - Φ_i: Phase of structure i
    - D_i: Dissonance measure |Φ_i - Ω|
    - ∇_Φ D_i: Gradient of dissonance w.r.t. phase
    - Transmute(D_i): Cross-monad resonance mediation term
    - Grace(Φ_i): Grace field contribution
    """

    alpha_i: float  # Resonance gradient coefficient
    beta_i: float   # Transmutation coefficient
    gamma_i: float  # Grace field coefficient
    dt: float = 0.01  # Time step for evolution
    max_iterations: int = 1000
    convergence_threshold: float = 1e-6
    oscillation_threshold: float = 0.1

    def __post_init__(self):
        """Validate coefficients."""
        if not all(c >= 0 for c in [self.alpha_i, self.beta_i, self.gamma_i]):
            raise ValueError("All evolution coefficients must be non-negative")

    def compute_dissonance_gradient(self, phi_i: float, omega: OmegaSignature,
                                 structure: ObjectG) -> float:
        """Compute ∇_Φ D_i - gradient of dissonance with respect to phase.

        Theory: Dissonance D_i = |Φ_i - Ω| where Ω is the attractor phase.
        The gradient ∇_Φ D_i = sign(Φ_i - Ω) for the phase component.
        """
        # Get current structure resonance (which measures dissonance)
        current_resonance = compute_resonance_alignment(structure, omega)

        # For gradient computation, we need to consider how resonance changes
        # with phase perturbations. This is a simplified gradient estimate.
        dissonance = 1.0 - current_resonance  # Convert resonance to dissonance

        # Estimate gradient by perturbing phase slightly
        epsilon = 1e-6
        perturbed_structure = self._perturb_phase(structure, epsilon)
        perturbed_resonance = compute_resonance_alignment(perturbed_structure, omega)
        perturbed_dissonance = 1.0 - perturbed_resonance

        gradient = (perturbed_dissonance - dissonance) / epsilon
        return gradient

    def compute_transmutation(self, phi_i: float, omega: OmegaSignature,
                            structure: ObjectG) -> float:
        """Compute Transmute(D_i) - cross-monad resonance mediation.

        This implements the transmutative mediation between different
        monadic structures to enhance overall coherence.
        """
        # For now, implement a simplified transmutation based on
        # structure connectivity and phase diversity
        coherence = compute_coherence(structure)

        # Transmutation factor based on phase diversity within structure
        phase_diversity = self._compute_phase_diversity(structure)

        # Transmutation promotes structures with balanced phase diversity
        # Optimal diversity around 0.5-0.7 for good mediation
        optimal_diversity = 0.6
        transmutation_factor = 1.0 - abs(phase_diversity - optimal_diversity)

        return coherence * transmutation_factor

    def compute_grace_field(self, phi_i: float, field_regime: FieldRegime) -> float:
        """Compute Grace(Φ_i) - grace field contribution to evolution.

        The grace field provides the potential landscape that guides
        structures toward more coherent configurations.
        """
        # Grace field strength varies by regime
        regime_multipliers = {
            FieldRegime.NON_BEING: 0.1,
            FieldRegime.VACUUM: 0.3,
            FieldRegime.DARK_SECTOR: 0.6,
            FieldRegime.MATTER: 0.8,
            FieldRegime.OMEGA: 1.0
        }

        base_grace = regime_multipliers.get(field_regime, 0.5)

        # Grace field also depends on current phase coherence
        # More coherent phases receive stronger grace field guidance
        # Canonical baseline: φ⁻¹ ≈ 0.618 represents natural vacuum potential
        PHI_INVERSE = 1 / 1.618033988749  # ≈ 0.618
        phase_coherence_factor = min(1.0, phi_i / PHI_INVERSE)  # Normalize relative to golden baseline

        return base_grace * phase_coherence_factor

    def evolve_structure(self, structure: ObjectG, omega: OmegaSignature,
                        field_regime: FieldRegime,
                        max_steps: int = None) -> Tuple[ObjectG, EvolutionMetrics]:
        """Evolve a structure according to the dynamic equation.

        Returns the evolved structure and evolution metrics.
        """
        current_structure = structure.copy() if hasattr(structure, 'copy') else structure
        metrics = EvolutionMetrics()
        max_steps = max_steps or self.max_iterations

        for step in range(max_steps):
            # Compute current state
            current_coherence = compute_coherence(current_structure)
            current_resonance = compute_resonance_alignment(current_structure, omega)

            # Extract representative phase for evolution (simplified)
            phi_i = self._extract_representative_phase(current_structure)

            # Compute evolution terms
            dissonance_gradient = self.compute_dissonance_gradient(
                phi_i, omega, current_structure
            )
            transmutation_term = self.compute_transmutation(
                phi_i, omega, current_structure
            )
            grace_term = self.compute_grace_field(phi_i, field_regime)

            # Apply evolution equation: dΦ_i/dt = -α ∇_Φ D_i + β Transmute + γ Grace
            phase_derivative = (
                -self.alpha_i * dissonance_gradient +
                self.beta_i * transmutation_term +
                self.gamma_i * grace_term
            )

            # Update phase (simplified - in full implementation would update node phases)
            new_phi_i = phi_i + phase_derivative * self.dt

            # Apply phase update to structure
            evolved_structure = self._apply_phase_evolution(
                current_structure, phi_i, new_phi_i, omega
            )

            # Update metrics
            evolved_coherence = compute_coherence(evolved_structure)
            evolved_resonance = compute_resonance_alignment(evolved_structure, omega)

            metrics.coherence_history.append(evolved_coherence)
            metrics.resonance_history.append(evolved_resonance)
            metrics.phase_history.append({'step': step, 'phase': new_phi_i})

            # Check convergence
            if self._check_convergence(metrics, step):
                metrics.state = EvolutionState.CONVERGING
                break

            # Check oscillation
            if self._check_oscillation(metrics, step):
                metrics.state = EvolutionState.OSCILLATING
                break

            current_structure = evolved_structure

            # Safety check for divergence
            if self._check_divergence(metrics, step):
                metrics.state = EvolutionState.DIVERGING
                break

        # Final metrics computation
        self._compute_final_metrics(metrics)

        return evolved_structure, metrics

    def _perturb_phase(self, structure: ObjectG, epsilon: float) -> ObjectG:
        """Create a perturbed version of the structure for gradient computation."""
        # Simplified perturbation - in full implementation would modify node phases
        perturbed_structure = ObjectG(
            nodes=structure.nodes.copy(),
            edges=structure.edges.copy(),
            labels=structure.labels.copy()
        )

        # Perturb a representative node's phase
        if perturbed_structure.labels:
            first_node_id = next(iter(perturbed_structure.labels.keys()))
            original_label = perturbed_structure.labels[first_node_id]

            # Add small phase perturbation
            new_numer = original_label.phase_numer + epsilon
            new_denom = original_label.phase_denom

            perturbed_label = NodeLabel(
                original_label.kind,
                int(new_numer),
                new_denom,
                original_label.monadic_id
            )

            perturbed_structure.labels[first_node_id] = perturbed_label

        return perturbed_structure

    def _compute_phase_diversity(self, structure: ObjectG) -> float:
        """Compute phase diversity within a structure."""
        if not structure.labels:
            return 0.0

        phases = []
        for label in structure.labels.values():
            phase_angle = 2 * math.pi * label.phase_numer / label.phase_denom
            phases.append(phase_angle)

        if len(phases) <= 1:
            return 0.0

        # Compute standard deviation of phases normalized to [0,1]
        mean_phase = sum(phases) / len(phases)
        variance = sum((p - mean_phase) ** 2 for p in phases) / len(phases)
        std_dev = math.sqrt(variance)

        # Normalize to [0,1] range (max possible std dev is π)
        return min(1.0, std_dev / math.pi)

    def _extract_representative_phase(self, structure: ObjectG) -> float:
        """Extract a representative phase value from the structure."""
        if not structure.labels:
            return 0.0

        # Use average phase as representative
        total_phase = 0.0
        count = 0

        for label in structure.labels.values():
            phase_angle = 2 * math.pi * label.phase_numer / label.phase_denom
            total_phase += phase_angle
            count += 1

        return total_phase / count if count > 0 else 0.0

    def _apply_phase_evolution(self, structure: ObjectG, old_phi: float,
                             new_phi: float, omega: OmegaSignature) -> ObjectG:
        """Apply phase evolution to structure nodes."""
        evolved_structure = ObjectG(
            nodes=structure.nodes.copy(),
            edges=structure.edges.copy(),
            labels=structure.labels.copy()
        )

        # Apply phase evolution to all nodes (simplified approach)
        phase_shift = new_phi - old_phi

        for node_id, label in evolved_structure.labels.items():
            # Convert phase shift back to rational representation
            old_phase_angle = 2 * math.pi * label.phase_numer / label.phase_denom
            new_phase_angle = old_phase_angle + phase_shift

            # Clamp to valid range [0, 2π)
            new_phase_angle = new_phase_angle % (2 * math.pi)

            # Convert back to rational (simplified)
            new_numer = int(round(new_phase_angle * label.phase_denom / (2 * math.pi)))
            new_numer = max(0, min(new_numer, 2 * label.phase_denom - 1))  # Ensure valid range

            evolved_label = NodeLabel(
                label.kind,
                new_numer,
                label.phase_denom,
                label.monadic_id
            )

            evolved_structure.labels[node_id] = evolved_label

        return evolved_structure

    def _check_convergence(self, metrics: EvolutionMetrics, step: int) -> bool:
        """Check if evolution is converging."""
        if len(metrics.coherence_history) < 10:
            return False

        recent_coherence = metrics.coherence_history[-10:]
        coherence_range = max(recent_coherence) - min(recent_coherence)

        return coherence_range < self.convergence_threshold

    def _check_oscillation(self, metrics: EvolutionMetrics, step: int) -> bool:
        """Check if evolution is oscillating."""
        if len(metrics.coherence_history) < 20:
            return False

        recent_coherence = metrics.coherence_history[-20:]
        oscillations = 0

        for i in range(1, len(recent_coherence) - 1):
            if (recent_coherence[i] - recent_coherence[i-1]) * (recent_coherence[i+1] - recent_coherence[i]) < 0:
                oscillations += 1

        oscillation_ratio = oscillations / (len(recent_coherence) - 2)
        return oscillation_ratio > 0.3  # More than 30% of steps are oscillation points

    def _check_divergence(self, metrics: EvolutionMetrics, step: int) -> bool:
        """Check if evolution is diverging."""
        if len(metrics.coherence_history) < 10:
            return False

        # Check if coherence is consistently decreasing over a longer window
        recent_coherence = metrics.coherence_history[-10:]

        # Only consider it divergence if coherence is decreasing significantly
        # and consistently over multiple steps
        coherence_changes = [
            recent_coherence[i+1] - recent_coherence[i]
            for i in range(len(recent_coherence) - 1)
        ]

        negative_changes = [c for c in coherence_changes if c < -0.01]  # Small threshold
        negative_ratio = len(negative_changes) / len(coherence_changes)

        # Divergence if more than 70% of recent changes are negative
        # and the trend is significantly downward
        trend = sum(coherence_changes) / len(coherence_changes)
        return negative_ratio > 0.7 and trend < -0.05

    def _compute_final_metrics(self, metrics: EvolutionMetrics):
        """Compute final evolution metrics."""
        if len(metrics.coherence_history) < 2:
            return

        # Evolution rate (change per step)
        coherence_changes = [
            metrics.coherence_history[i+1] - metrics.coherence_history[i]
            for i in range(len(metrics.coherence_history) - 1)
        ]
        metrics.evolution_rate = sum(coherence_changes) / len(coherence_changes)

        # Convergence rate (how quickly approaching final state)
        if len(metrics.coherence_history) >= 10:
            final_coherence = metrics.coherence_history[-1]
            initial_coherence = metrics.coherence_history[0]

            if abs(final_coherence - initial_coherence) > 1e-6:
                total_change = final_coherence - initial_coherence
                steps_needed = len(metrics.coherence_history)
                metrics.convergence_rate = total_change / steps_needed

        # Oscillation amplitude
        if len(metrics.coherence_history) >= 20:
            recent_coherence = metrics.coherence_history[-20:]
            metrics.oscillation_amplitude = max(recent_coherence) - min(recent_coherence)


@dataclass
class ModeCoefficients:
    """Dynamic mode coefficient management based on field regime."""

    coefficient_sets: Dict[FieldRegime, Tuple[float, float, float]] = field(default_factory=lambda: {
        FieldRegime.NON_BEING: (0.1, 0.1, 0.1),
        FieldRegime.VACUUM: (0.3, 0.2, 0.4),
        FieldRegime.DARK_SECTOR: (0.5, 0.4, 0.6),
        FieldRegime.MATTER: (0.7, 0.6, 0.8),
        FieldRegime.OMEGA: (1.0, 1.0, 1.0)
    })

    def get_coefficients(self, field_regime: FieldRegime) -> Tuple[float, float, float]:
        """Get evolution coefficients for a field regime."""
        return self.coefficient_sets.get(field_regime, (0.5, 0.5, 0.5))

    def adapt_coefficients(self, field_regime: FieldRegime,
                          evolution_metrics: EvolutionMetrics) -> Tuple[float, float, float]:
        """Adapt coefficients based on evolution performance."""
        base_coeffs = self.get_coefficients(field_regime)

        # Adjust based on evolution state
        if evolution_metrics.state == EvolutionState.DIVERGING:
            # Reduce coefficients to slow down evolution
            adaptation_factor = 0.7
        elif evolution_metrics.state == EvolutionState.OSCILLATING:
            # Reduce alpha (resonance gradient) to dampen oscillations
            alpha_factor = 0.6
            return (base_coeffs[0] * alpha_factor, base_coeffs[1], base_coeffs[2])
        elif evolution_metrics.state == EvolutionState.CONVERGING:
            # Slightly increase coefficients for faster convergence
            adaptation_factor = 1.2
        else:
            adaptation_factor = 1.0

        return tuple(c * adaptation_factor for c in base_coeffs)


# Factory functions for easy creation
def create_dynamic_evolution(alpha: float = 0.5, beta: float = 0.3,
                           gamma: float = 0.4, dt: float = 0.01) -> DynamicPhaseEvolution:
    """Create a dynamic evolution system with specified coefficients."""
    return DynamicPhaseEvolution(alpha_i=alpha, beta_i=beta, gamma_i=gamma, dt=dt)


def create_mode_coefficients() -> ModeCoefficients:
    """Create default mode coefficient manager."""
    return ModeCoefficients()


__all__ = [
    "DynamicPhaseEvolution",
    "ModeCoefficients",
    "EvolutionState",
    "EvolutionMetrics",
    "create_dynamic_evolution",
    "create_mode_coefficients",
]

File: FIRM-Core/FIRM_ui/FIRM_dsl/dynamic_evolution.js
/**
 * dynamic_evolution.js
 *
 * JavaScript implementation of dynamic phase evolution for Soul Garbage Collection.
 *
 * Implements the theoretical equation:
 * dΦ_i/dt = -α_i ∇_Φ D_i + β_i Transmute(D_i) + γ_i Grace(Φ_i)
 *
 * This bridges the gap between static analysis and dynamic evolution in the 𝒮-GC system.
 */

// Import existing DSL modules
import { ObjectG } from './core.js';
import { computeResonanceAlignment, deriveOmegaSignature } from './resonance.js';
import { compute_coherence } from './coherence.js';
import { FieldRegime } from './soul_garbage_collection.js';

// Evolution states enum
export const EvolutionState = {
    STABLE: 'stable',
    EVOLVING: 'evolving',
    CONVERGING: 'converging',
    DIVERGING: 'diverging',
    OSCILLATING: 'oscillating'
};

// Evolution metrics class
export class EvolutionMetrics {
    constructor() {
        this.coherence_history = [];
        this.resonance_history = [];
        this.phase_history = [];
        this.evolution_rate = 0.0;
        this.convergence_rate = 0.0;
        this.oscillation_amplitude = 0.0;
        this.state = EvolutionState.STABLE;
    }
}

// Dynamic phase evolution class
export class DynamicPhaseEvolution {
    constructor(alpha_i = 0.5, beta_i = 0.3, gamma_i = 0.4, dt = 0.01) {
        this.alpha_i = alpha_i;  // Resonance gradient coefficient
        this.beta_i = beta_i;    // Transmutation coefficient
        this.gamma_i = gamma_i;  // Grace field coefficient
        this.dt = dt;            // Time step for evolution
        this.max_iterations = 1000;
        this.convergence_threshold = 1e-6;
        this.oscillation_threshold = 0.1;

        // Validate coefficients
        if (![this.alpha_i, this.beta_i, this.gamma_i].every(c => c >= 0)) {
            throw new Error("All evolution coefficients must be non-negative");
        }
    }

    computeDissonanceGradient(phi_i, omega, structure) {
        // Get current structure resonance (which measures dissonance)
        const currentResonance = computeResonanceAlignment(structure, omega);
        const dissonance = 1.0 - currentResonance;

        // Estimate gradient by perturbing phase slightly
        const epsilon = 1e-6;
        const perturbedStructure = this._perturbPhase(structure, epsilon);
        const perturbedResonance = computeResonanceAlignment(perturbedStructure, omega);
        const perturbedDissonance = 1.0 - perturbedResonance;

        const gradient = (perturbedDissonance - dissonance) / epsilon;
        return gradient;
    }

    computeTransmutation(phi_i, omega, structure) {
        // Simplified transmutation based on structure connectivity and phase diversity
        const coherence = compute_coherence(structure);

        // Transmutation factor based on phase diversity within structure
        const phaseDiversity = this._computePhaseDiversity(structure);

        // Transmutation promotes structures with balanced phase diversity
        // Optimal diversity around 0.5-0.7 for good mediation
        const optimalDiversity = 0.6;
        const transmutationFactor = 1.0 - Math.abs(phaseDiversity - optimalDiversity);

        return coherence * transmutationFactor;
    }

    computeGraceField(phi_i, fieldRegime) {
        // Grace field strength varies by regime
        const regimeMultipliers = {
            [FieldRegime.NON_BEING]: 0.1,
            [FieldRegime.VACUUM]: 0.3,
            [FieldRegime.DARK_SECTOR]: 0.6,
            [FieldRegime.MATTER]: 0.8,
            [FieldRegime.OMEGA]: 1.0
        };

        const baseGrace = regimeMultipliers[fieldRegime] || 0.5;

        // Grace field also depends on current phase coherence
        // More coherent phases receive stronger grace field guidance
        // Canonical baseline: φ⁻¹ ≈ 0.618 represents natural vacuum potential
        const PHI_INVERSE = 1 / 1.618033988749;  // ≈ 0.618
        const phaseCoherenceFactor = Math.min(1.0, phi_i / PHI_INVERSE);

        return baseGrace * phaseCoherenceFactor;
    }

    evolveStructure(structure, omega, fieldRegime, maxSteps = null) {
        // Create a copy of the structure for evolution
        let currentStructure = new ObjectG({
            nodes: [...structure.nodes],
            edges: [...structure.edges],
            labels: {...structure.labels}
        });

        const metrics = new EvolutionMetrics();
        const maxIterations = maxSteps || this.max_iterations;

        for (let step = 0; step < maxIterations; step++) {
            // Compute current state
            const currentCoherence = compute_coherence(currentStructure);
            const currentResonance = computeResonanceAlignment(currentStructure, omega);

            // Extract representative phase for evolution (simplified)
            const phi_i = this._extractRepresentativePhase(currentStructure);

            // Compute evolution terms
            const dissonanceGradient = this.computeDissonanceGradient(phi_i, omega, currentStructure);
            const transmutationTerm = this.computeTransmutation(phi_i, omega, currentStructure);
            const graceTerm = this.computeGraceField(phi_i, fieldRegime);

            // Apply evolution equation: dΦ_i/dt = -α ∇_Φ D_i + β Transmute + γ Grace
            const phaseDerivative = (
                -this.alpha_i * dissonanceGradient +
                this.beta_i * transmutationTerm +
                this.gamma_i * graceTerm
            );

            // Update phase
            const newPhi_i = phi_i + phaseDerivative * this.dt;

            // Apply phase update to structure
            const evolvedStructure = this._applyPhaseEvolution(currentStructure, phi_i, newPhi_i, omega);

            // Update metrics
            const evolvedCoherence = compute_coherence(evolvedStructure);
            const evolvedResonance = computeResonanceAlignment(evolvedStructure, omega);

            metrics.coherence_history.push(evolvedCoherence);
            metrics.resonance_history.push(evolvedResonance);
            metrics.phase_history.push({ step, phase: newPhi_i });

            // Check convergence
            if (this._checkConvergence(metrics, step)) {
                metrics.state = EvolutionState.CONVERGING;
                break;
            }

            // Check oscillation
            if (this._checkOscillation(metrics, step)) {
                metrics.state = EvolutionState.OSCILLATING;
                break;
            }

            currentStructure = evolvedStructure;

            // Safety check for divergence
            if (this._checkDivergence(metrics, step)) {
                metrics.state = EvolutionState.DIVERGING;
                break;
            }
        }

        // Final metrics computation
        this._computeFinalMetrics(metrics);

        return { evolvedStructure: currentStructure, metrics };
    }

    _perturbPhase(structure, epsilon) {
        // Create a perturbed version of the structure for gradient computation
        const perturbedStructure = new ObjectG({
            nodes: [...structure.nodes],
            edges: [...structure.edges],
            labels: {...structure.labels}
        });

        // Perturb a representative node's phase
        if (perturbedStructure.labels && Object.keys(perturbedStructure.labels).length > 0) {
            const firstNodeId = Object.keys(perturbedStructure.labels)[0];
            const originalLabel = perturbedStructure.labels[firstNodeId];

            // Add small phase perturbation
            const newNumer = originalLabel.phase_numer + epsilon;
            const newDenom = originalLabel.phase_denom;

            perturbedStructure.labels[firstNodeId] = {
                kind: originalLabel.kind,
                phase_numer: Math.round(newNumer),
                phase_denom: newDenom,
                monadic_id: originalLabel.monadic_id
            };
        }

        return perturbedStructure;
    }

    _computePhaseDiversity(structure) {
        if (!structure.labels || Object.keys(structure.labels).length === 0) {
            return 0.0;
        }

        const phases = [];
        for (const label of Object.values(structure.labels)) {
            const phaseAngle = 2 * Math.PI * label.phase_numer / label.phase_denom;
            phases.push(phaseAngle);
        }

        if (phases.length <= 1) {
            return 0.0;
        }

        // Compute standard deviation of phases normalized to [0,1]
        const meanPhase = phases.reduce((sum, p) => sum + p, 0) / phases.length;
        const variance = phases.reduce((sum, p) => sum + Math.pow(p - meanPhase, 2), 0) / phases.length;
        const stdDev = Math.sqrt(variance);

        // Normalize to [0,1] range (max possible std dev is π)
        return Math.min(1.0, stdDev / Math.PI);
    }

    _extractRepresentativePhase(structure) {
        if (!structure.labels || Object.keys(structure.labels).length === 0) {
            return 0.0;
        }

        // Use average phase as representative
        let totalPhase = 0.0;
        let count = 0;

        for (const label of Object.values(structure.labels)) {
            const phaseAngle = 2 * Math.PI * label.phase_numer / label.phase_denom;
            totalPhase += phaseAngle;
            count++;
        }

        return count > 0 ? totalPhase / count : 0.0;
    }

    _applyPhaseEvolution(structure, oldPhi, newPhi, omega) {
        const evolvedStructure = new ObjectG({
            nodes: [...structure.nodes],
            edges: [...structure.edges],
            labels: {...structure.labels}
        });

        // Apply phase evolution to all nodes (simplified approach)
        const phaseShift = newPhi - oldPhi;

        for (const [nodeId, label] of Object.entries(evolvedStructure.labels)) {
            // Convert phase shift back to rational representation
            const oldPhaseAngle = 2 * Math.PI * label.phase_numer / label.phase_denom;
            const newPhaseAngle = oldPhaseAngle + phaseShift;

            // Convert back to rational (simplified)
            const newNumer = Math.round(newPhaseAngle * label.phase_denom / (2 * Math.PI));
            const clampedNumer = newNumer % (2 * label.phase_denom);

            evolvedStructure.labels[nodeId] = {
                kind: label.kind,
                phase_numer: clampedNumer,
                phase_denom: label.phase_denom,
                monadic_id: label.monadic_id
            };
        }

        return evolvedStructure;
    }

    _checkConvergence(metrics, step) {
        if (metrics.coherence_history.length < 10) {
            return false;
        }

        const recentCoherence = metrics.coherence_history.slice(-10);
        const coherenceRange = Math.max(...recentCoherence) - Math.min(...recentCoherence);

        return coherenceRange < this.convergence_threshold;
    }

    _checkOscillation(metrics, step) {
        if (metrics.coherence_history.length < 20) {
            return false;
        }

        const recentCoherence = metrics.coherence_history.slice(-20);
        let oscillations = 0;

        for (let i = 1; i < recentCoherence.length - 1; i++) {
            if ((recentCoherence[i] - recentCoherence[i-1]) * (recentCoherence[i+1] - recentCoherence[i]) < 0) {
                oscillations++;
            }
        }

        const oscillationRatio = oscillations / (recentCoherence.length - 2);
        return oscillationRatio > 0.3; // More than 30% of steps are oscillation points
    }

    _checkDivergence(metrics, step) {
        if (metrics.coherence_history.length < 5) {
            return false;
        }

        // Check if coherence is consistently decreasing or becoming erratic
        const recentCoherence = metrics.coherence_history.slice(-5);
        if (recentCoherence.every(c => c < 0.1)) {
            return true;
        }

        // Check for rapid degradation
        const coherenceTrend = recentCoherence[recentCoherence.length - 1] - recentCoherence[0];
        return coherenceTrend < -0.5; // Rapid coherence loss
    }

    _computeFinalMetrics(metrics) {
        if (metrics.coherence_history.length < 2) {
            return;
        }

        // Evolution rate (change per step)
        const coherenceChanges = [];
        for (let i = 0; i < metrics.coherence_history.length - 1; i++) {
            coherenceChanges.push(metrics.coherence_history[i+1] - metrics.coherence_history[i]);
        }
        metrics.evolution_rate = coherenceChanges.reduce((sum, change) => sum + change, 0) / coherenceChanges.length;

        // Convergence rate (how quickly approaching final state)
        if (metrics.coherence_history.length >= 10) {
            const finalCoherence = metrics.coherence_history[metrics.coherence_history.length - 1];
            const initialCoherence = metrics.coherence_history[0];

            if (Math.abs(finalCoherence - initialCoherence) > 1e-6) {
                const totalChange = finalCoherence - initialCoherence;
                const stepsNeeded = metrics.coherence_history.length;
                metrics.convergence_rate = totalChange / stepsNeeded;
            }
        }

        // Oscillation amplitude
        if (metrics.coherence_history.length >= 20) {
            const recentCoherence = metrics.coherence_history.slice(-20);
            metrics.oscillation_amplitude = Math.max(...recentCoherence) - Math.min(...recentCoherence);
        }
    }
}

// Mode coefficients management
export class ModeCoefficients {
    constructor() {
        this.coefficient_sets = {
            [FieldRegime.NON_BEING]: [0.1, 0.1, 0.1],
            [FieldRegime.VACUUM]: [0.3, 0.2, 0.4],
            [FieldRegime.DARK_SECTOR]: [0.5, 0.4, 0.6],
            [FieldRegime.MATTER]: [0.7, 0.6, 0.8],
            [FieldRegime.OMEGA]: [1.0, 1.0, 1.0]
        };
    }

    getCoefficients(fieldRegime) {
        return this.coefficient_sets[fieldRegime] || [0.5, 0.5, 0.5];
    }

    adaptCoefficients(fieldRegime, evolutionMetrics) {
        const baseCoeffs = this.getCoefficients(fieldRegime);

        // Adjust based on evolution state
        if (evolutionMetrics.state === EvolutionState.DIVERGING) {
            // Reduce coefficients to slow down evolution
            const adaptationFactor = 0.7;
            return baseCoeffs.map(c => c * adaptationFactor);
        } else if (evolutionMetrics.state === EvolutionState.OSCILLATING) {
            // Reduce alpha (resonance gradient) to dampen oscillations
            const alphaFactor = 0.6;
            return [baseCoeffs[0] * alphaFactor, baseCoeffs[1], baseCoeffs[2]];
        } else if (evolutionMetrics.state === EvolutionState.CONVERGING) {
            // Slightly increase coefficients for faster convergence
            const adaptationFactor = 1.2;
            return baseCoeffs.map(c => c * adaptationFactor);
        } else {
            return baseCoeffs;
        }
    }
}

// Factory functions
export function createDynamicEvolution(alpha = 0.5, beta = 0.3, gamma = 0.4, dt = 0.01) {
    return new DynamicPhaseEvolution(alpha, beta, gamma, dt);
}

export function createModeCoefficients() {
    return new ModeCoefficients();
}
